### PR TITLE
Tiered index info MOD-4315

### PR DESCRIPTION
**Pull request by DvirDukhan**
**Describe the changes in the pull request**

This PR implements the `info` and `infoIterator` capabilities for tiered index
**Which issues this PR fixes**
MOD-4967

**Main objects this PR modified**
`VecSimIndexInfo` now contains `commonInfo` section and each specilized index holds its own specielized info.

**Mark if applicable**

- [x] This PR introduces API changes - Info struct has been changed
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/unit/test_utils.cpp [hunk 1]
```diff
@@ -109,126 +109,154 @@ void runBatchIteratorSearchTest(VecSimBatchIterator *batch_iterator, size_t n_re
...
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoField->fieldValue.uintegerValue, info.commonInfo.blockSize);
+        } else if (!strcmp(infoField->fieldName, VecSimCommonStrings::MEMORY_STRING)) {
             // Memory.
-            ASSERT_EQ(infoFiled->fieldType, INFOFIELD_UINT64);
-            ASSERT_EQ(infoFiled->fieldValue.uintegerValue, info.bfInfo.memory);
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoField->fieldValue.uintegerValue, info.commonInfo.memory);
         } else {
             ASSERT_TRUE(false);
```
Comment (GuyAv46) [inline tests/unit/test_utils.cpp:183]:
```suggestion
            FAIL();
```

File: tests/unit/test_utils.cpp [hunk 2]
```diff
@@ -109,126 +109,154 @@ void runBatchIteratorSearchTest(VecSimBatchIterator *batch_iterator, size_t n_re
...
-            ASSERT_EQ(infoFiled->fieldType, INFOFIELD_UINT64);
-            ASSERT_EQ(infoFiled->fieldValue.uintegerValue, info.hnswInfo.memory);
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoField->fieldValue.uintegerValue, info.commonInfo.memory);
+        } else if (!strcmp(infoField->fieldName, VecSimCommonStrings::BLOCK_SIZE_STRING)) {
+            // Block size.
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoField->fieldValue.uintegerValue, info.commonInfo.blockSize);
         } else {
             ASSERT_TRUE(false);
```
Comment (GuyAv46) [inline tests/unit/test_utils.cpp:261]:
```suggestion
            FAIL();
```

File: tests/unit/test_utils.cpp [hunk 3]
```diff
@@ -109,126 +109,154 @@ void runBatchIteratorSearchTest(VecSimBatchIterator *batch_iterator, size_t n_re
...
     }
 }
 
 void compareHNSWIndexInfoToIterator(VecSimIndexInfo info, VecSimInfoIterator *infoIter) {
-    ASSERT_EQ(15, VecSimInfoIterator_NumberOfFields(infoIter));
+    ASSERT_EQ(16, VecSimInfoIterator_NumberOfFields(infoIter));
     while (VecSimInfoIterator_HasNextField(infoIter)) {
-        VecSim_InfoField *infoFiled = VecSimInfoIterator_NextField(infoIter);
-        if (!strcmp(infoFiled->fieldName, VecSimCommonStrings::ALGORITHM_STRING)) {
+        VecSim_InfoField *infoField = VecSimInfoIterator_NextField(infoIter);
```
Comment (GuyAv46) [inline tests/unit/test_utils.cpp:191]: nice catch

File: tests/unit/test_hnsw_tiered.cpp [hunk 4]
```diff
@@ -2799,3 +2798,201 @@ TYPED_TEST(HNSWTieredIndexTestBasic, overwriteVectorAsync) {
...
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_UINT64);
+            ASSERT_EQ(infoField->fieldValue.uintegerValue, info.tieredInfo.backgroundIndexing);
+        } else if (!strcmp(infoField->fieldName, VecSimCommonStrings::FRONTEND_INDEX_STRING)) {
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_ITERATOR);
+            compareFlatIndexInfoToIterator(frontendIndexInfo, infoField->fieldValue.iteratorValue);
+        } else if (!strcmp(infoField->fieldName, VecSimCommonStrings::BACKEND_INDEX_STRING)) {
+            ASSERT_EQ(infoField->fieldType, INFOFIELD_ITERATOR);
+            compareHNSWIndexInfoToIterator(backendIndexInfo, infoField->fieldValue.iteratorValue);
+        } else {
+            assert(false);
```
Comment (GuyAv46) [inline tests/unit/test_hnsw_tiered.cpp:2993]:
```suggestion
            FAIL();
```

**Review by GuyAv46 — COMMENTED** [review]
General comment:
How does the memory is being reported in tiered? 3 parts? Or is there also the sum of the parts?

File: src/VecSim/vec_sim_tiered_index.h [hunk 5]
```diff
@@ -159,6 +168,80 @@ VecSimTieredIndex<DataType, DistType>::topKQuery(const void *queryBlob, size_t k
...
+    case VecSimAlgo_TIERED:
+        assert(false && "Invalid backend algorithm");
+    }
+
+    info.tieredInfo.backendCommonInfo = backendInfo.commonInfo;
+    // For now, this is hard coded to FLAT
+    info.tieredInfo.frontendCommonInfo = frontendInfo.commonInfo;
+    info.tieredInfo.bfInfo = frontendInfo.bfInfo;
+
+    info.tieredInfo.backgroundIndexing = this->frontendIndex->indexSize() > 0;
```
Comment (alonre24) [inline src/VecSim/vec_sim_tiered_index.h:203]: I think we should expose the number of zombies in the tiered index info section as well (num marked deleted)

**Pull request merged**